### PR TITLE
GPC-347 step function xml to dynamodb

### PR DESCRIPTION
**Pull request by CarlyG55**

### CODE REVIEW

File: terraform/modules/gro_ingestion_service/lambdas/src/functions/mapXml.function.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+const pushRecord = async (record: PutItemCommandInput) => {
+    const command = new PutItemCommand(record)
+    await client.send(command)
+}
+
+const generateRecord = (deathRegistration: GroDeathRegistration): PutItemCommandInput => {
+    const eventHash = hash(deathRegistration)
+    return {
+        Item: mapToDynamoDbItem(deathRegistration, eventHash),
```
Comment (oskarwilliams378) [inline terraform/modules/gro_ingestion_service/lambdas/src/functions/mapXml.function.ts:19]: Could we do the hashing in the mapping?

File: terraform/modules/gro_ingestion_service/lambdas/src/models/GroDeathRegistration.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    DeceasedAddress: {
+        Flat: string,
+        Building: string,
+        Line: string[],
+        Postcode: string,
+    },
+}
+
+
+export const mapToDynamoDbItem = (deathRecord: GroDeathRegistration, hash: string): { [key: string]: AttributeValue } => {
```
Comment (oskarwilliams378) [inline terraform/modules/gro_ingestion_service/lambdas/src/models/GroDeathRegistration.ts:48]: I would rather we map to EventRecord, then we can run `marshal(eventRecord)` to get the dynamoDbItem. marshal is in `@aws-sdk/util-dynamodb`

File: terraform/modules/gro_ingestion_service/lambdas/src/models/EventRecord.ts [hunk 3]
```diff
@@ -43,3 +44,28 @@ export const mapToEventRecord = (image: { [key: string]: AttributeValue }): Even
         Postcode: image["Postcode"].S ?? "",
     }
 }
+
```
Comment (oskarwilliams378) [inline terraform/modules/gro_ingestion_service/lambdas/src/models/EventRecord.ts:47]: for the one above here we can do `unmarshal(image) as EventRecord` as using `@aws-sdk/util-dynamodb`

File: terraform/modules/gro_ingestion_service/lambdas/src/functions/splitXml.function.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        Bucket: event.Records[0].s3.bucket.name,
+        Key: event.Records[0].s3.object.key
+    })
+    const groFileResponse = await client.send(getGroCommand)
+
+    return groFileResponse.Body?.transformToString()
+}
+
+const parseGroXml = (groXml: string): GroXmlFile => {
+    const parser = new XMLParser()
```
Comment (oskarwilliams378) [inline terraform/modules/gro_ingestion_service/lambdas/src/functions/splitXml.function.ts:19]: Can we create this and the S3 client outside of the functions

File: terraform/modules/gro_ingestion_service/lambdas/src/functions/splitXml.function.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    return parser.parse(groXml)
+}
+
+export const handler: Handler = async (event: S3Event) => {
+    const groFile = await getGroFile(event)
```
Comment (oskarwilliams378) [inline terraform/modules/gro_ingestion_service/lambdas/src/functions/splitXml.function.ts:25]: if this is null or empty string, we should probably error

File: terraform/modules/gro_ingestion_service/lambdas/src/models/GroXmlFile.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { GroDeathRegistration } from "./GroDeathRegistration"
+
+export interface GroXmlFile {
```
Comment (oskarwilliams378) [inline terraform/modules/gro_ingestion_service/lambdas/src/models/GroXmlFile.ts:3]: This isn't the XML file, this is the JSON object of it, could you pls try a rename (naming is hard :( )

**Review by oskarwilliams378 — CHANGES_REQUESTED** [review]

File: terraform/modules/gro_ingestion_service/lambdas/src/models/GroXmlFile.ts (hunk 6 above)
Comment (CarlyG55) [reply terraform/modules/gro_ingestion_service/lambdas/src/models/GroXmlFile.ts:3]: It's not needed anymore whoop

File: terraform/modules/gro_ingestion_service/lambdas/src/functions/splitXml.function.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+            error: `File with key ${event.Records[0].s3.object.key} not found`,
+        }
+        console.error("Failed to insert records into DynamoDB", logParams)
+        return {
+            statusCode: 404,
+        }
+    }
+
+    const groJson = parser.parse(groXml)
+    return { deathRegistrations: groJson.deathRegistrationGroup }
```
Comment (CarlyG55) [inline terraform/modules/gro_ingestion_service/lambdas/src/functions/splitXml.function.ts:34]: Not sure if we now want to add a status code here?

Comment (oskarwilliams378) [reply terraform/modules/gro_ingestion_service/lambdas/src/functions/splitXml.function.ts:34]: This is all happening inside lambdas from S3 to DynamoDB, I don't think we need a status code

**Pull request merged**